### DISCUSSION TITLE
API - Getting Current Layer

### DISCUSSION THREAD

**Discussion by phinphan20** [Arksine/moonraker#441] [General]
Is there anywhere in the API to get the current layer being printed?

**Comment by phinphan20** [comment]
Never mind figured it out.

**Reply by albertsj1** [reply]
@phinphan20  What was the solution?  It looks like it should be included in the display_stats printer object; however, I'm getting null values.

`curl -s 'http://myprinter/printer/objects/query?print_stats'  | jq '.' `
```json
{
  "result": {
    "eventtime": 92767.326975521,
    "status": {
      "print_stats": {
        "filename": "jma-llama_5h32m_0.20mm_200C_PLA_ENDER3S1PRO.gcode",
        "total_duration": 7885.3678659480065,
        "print_duration": 7728.45696691201,
        "filament_used": 5526.163270000981,
        "state": "printing",
        "message": "",
        "info": {
          "total_layer": null,
          "current_layer": null
        }
      }
    }
  }
}
```

**Reply by phinphan20** [reply]
@albertsj1 it has been awhile since I messed with this code but I believe this is the code I used to create the current layer.

I get current_position from here:  'http://myprinter/printer/objects/query?gcode_move'

Layer height from here:  'http://myprinter/server/files/metadata?filename='.$filename; - by using the filename to get the details of the file.
`$layer_height = $response['result']['layer_height'];`

```
$current_position =  $response['result']['status']['gcode_move']['gcode_position']['2'];
$current_layer =  $current_position / $layer_height;
$layer_progress = "Layer: ".$current_layer." / ".$layer_count;
```

You have to calculate it, I could never find it in the actual API as it was being printed.

I hope that makes sense.

**Reply by albertsj1** [reply]
ah, ok.  Thank you.  Yes that makes sense.

I wonder why it's not being reported in the api like it's supposed to be, since there are fields for it.

**Reply by phinphan20** [reply]
Good question, I fought this for a good while until I finally got it.

**Reply by Arksine** [reply]
Your slicer needs to be configured to emit the correct [gcode command](https://www.klipper3d.org/G-Codes.html#set_print_stats_info) for `print_stats` to report the total layer and current layer values.  Otherwise its necessary to estimate from the position.

**Reply by phinphan20** [reply]
Ok, I missed this when configuring the start gcode - I will set this and check it.  Thanks for this info.

**Comment by albertsj1** [comment]
ah, ok.  thank you very much.
